gen.c

```diff
@@ -0,0 +1,97 @@
+#include "gen.h"
+
+#include "brute.h"
+#include "single.h"
+#include <stdio.h>
+#include <string.h>
+
+status_t
+gen_context_init (gen_context_t *context, config_t *config, task_t *task)
+{
+  if (pthread_mutex_init (&context->mutex, NULL) != 0)
+    {
+      print_error ("Could not initialize a mutex\n");
+      return (S_FAILURE);
+    }
+
+  iter_state_init (&context->state, config->alph, task);
+
+  context->config = config;
+  context->cancelled = false;
+  context->password[0] = 0;
+
+  return (S_SUCCESS);
+}
+
+void *
+generator (void *context)
+{
+  gen_context_t *gen_context = (gen_context_t *)context;
+
+  task_t *task = gen_context->state.task; /* to reduce the amount of code */
+
+  task->from = 0;
+  task->to = gen_context->config->length;
+
+  st_context_t st_context = {
+    .hash = gen_context->config->hash,
+    .data = { .initialized = 0 },
+  };
+
+  while (!gen_context->cancelled && gen_context->password[0] == 0)
+    {
+      if (pthread_mutex_lock (&gen_context->mutex) != 0)
+        {
+          print_error ("Could not lock a mutex\n");
+          return (NULL);
+        }
+
+      gen_context->cancelled = !iter_state_next (&gen_context->state);
+
+      if (pthread_mutex_unlock (&gen_context->mutex) != 0)
+        {
+          print_error ("Could not unlock a mutex\n");
+          return (NULL);
+        }
+
+      task->to = task->from;
```

Comment:
You can't use shared task. You need to make a copy of current task under mutex.

Reply:
> You can't use shared task. You need to make a copy of current task under mutex.

Changed generator mechanics as you said.